staking/slash/double-sign.go

```diff
@@ -1,52 +1,385 @@
+							)
+							nowAmt.Sub(nowAmt, slashDebt)
+						}
+					}
+					// NOTE Assume did as much as could above, now check the undelegations
+					for _, undelegate := range delegationNow.Undelegations {
+						// the epoch matters, only those undelegation
+						// such that epoch>= doubleSignEpoch should be slashable
+						if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
+							if slashDebt.Cmp(common.Big0) <= 0 {
+								// paid off the slash debt
+								break
+							}
+
+							switch newBal := new(big.Int).Sub(
+								// My slash debt is 1.6 and my undelegate amount is 1.0
+								// so if (1.6 - 1.0) > 0
+								undelegate.Amount,
+								slashDebt); newBal.Cmp(common.Big0) {
+							case paidOffExact:
+								undelegate.Amount.SetInt64(0)
+							case haveEnoughToPayOff:
+								undelegate.Amount.Sub(undelegate.Amount, slashDebt)
+								mustStayAboveZeroAfterDebtReduce(
+									snapshot, current, slashDebt, slashDebt, slashTrack,
+								)
+								undelegate.Amount.Set(newBal)
+							case -1:
+								// But do have enough to pay off something at least
+								partialPayment := new(big.Int).Sub(slashDebt, new(big.Int).Abs(newBal))
+								mustStayAboveZeroAfterDebtReduce(
+									snapshot, current, slashDebt, partialPayment, slashTrack,
+								)
+								undelegate.Amount.SetInt64(0)
+							}
+						}
+					}
+					// NOTE at high slashing % rates, we could hit this situation
+					// because of the special casing of logic on min-self-delegation
+					if slashDebt.Cmp(common.Big0) == 1 {
+						mustStayAboveZeroAfterDebtReduce(
+							snapshot, current, slashDebt, slashDebt, slashTrack,
+						)
+					}
+				} else {
+					// NOTE everyone else, that is every plain delegation
+					// Current delegation has some money and slashdebt is still not paid off
+					// so contribute as much as can with current delegation amount
+					if nowAmt.Cmp(common.Big0) == 1 && slashDebt.Cmp(common.Big0) == 1 {
+						// 0.50_amount > 0.06_debt => slash == 0.0, nowAmt == 0.44
+						if nowAmt.Cmp(slashDebt) >= 0 {
+							nowAmt.Sub(nowAmt, slashDebt)
+							mustStayAboveZeroAfterDebtReduce(
+								snapshot, current, slashDebt, slashDebt, slashTrack,
+							)
+						} else {
+							// 0.50_amount < 2.4_debt =>, slash == 1.9, nowAmt == 0.0
+							mustStayAboveZeroAfterDebtReduce(
+								snapshot, current, slashDebt, nowAmt, slashTrack,
+							)
+							nowAmt.Sub(nowAmt, nowAmt)
+						}
+					}
+
+					// NOTE Assume did as much as could above, now check the undelegations
+					for _, undelegate := range delegationNow.Undelegations {
+						// the epoch matters, only those undelegation
+						// such that epoch>= doubleSignEpoch should be slashable
+						if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
+							if slashDebt.Cmp(common.Big0) <= 0 {
+								// paid off the slash debt
+								break
+							}
+
+							if undelegate.Amount.Cmp(common.Big0) == 1 &&
```

Comment:
The logic here is oversimplied. The newBal could be negative. Actually the one above for validator is correct.

Reply:
And I think we don't need to separate out validator/delegator slashing. After taking care of Banned field in sanity check, the two cases can be handled with the same logic.